README.md

```diff
@@ -9,21 +9,72 @@ rustup target add riscv64gc-unknown-none-elf
 cargo build
 ```
 
-## running
+## Running
 
-There is a `Makefile` provided with targets for running various hosts in qemu.
+There is a `Makefile` provided with targets for running various hosts in QEMU.
 
 ### Prerequisites
 
-- rust toolchain from [rustup](rustup.rs)
-- rustup target add riscv64gc-unknown-none-elf
-- install `gcc-riscv64-unknown-elf` for `riscv64-unknown-elf-objcopy`
-- qemu-system-riscv64 - build using  qemu
-  [instructions](https://wiki.qemu.org/Hosts/Linux) with
+Toolchains:
+- Rust: Install [rustup](rustup.rs), then: `rustup target add riscv64gc-unknown-none-elf`
+- GCC: Install `gcc-riscv64-unknown-elf`
+
+QEMU:
+- Out-of-tree patches are required; see table below.
+- Build using QEMU [instructions](https://wiki.qemu.org/Hosts/Linux) with
   `--target-list=riscv64-softmmu`
-  - Note qemu 7.0.0 has a bug emulating riscv and a newer version from git must be used.
 
-## Test VM
+OpenSBI:
+- Use source from upstream: https://github.com/riscv-software-src/opensbi
+- Build: `PLATFORM=generic CROSS_COMPILE=riscv64-unknown-linux-gnu- make`
+
+Linux kernel (for booting Linux VMs):
+- Out-of-tree patches are required; see table below.
+- Build: `ARCH=riscv CROSS_COMPILE=riscv64-unknown-linux-gnu- make defconfig Image`
+
+Latest known-working branches:
+
+| Project | Branch |
+| ------- | ------ |
+| QEMU    | https://github.com/rivosinc/qemu/tree/salus-integration-08172022 |
+| Linux   | https://github.com/rivosinc/linux/tree/salus-integration-08172022 |
+| OpenSBI | use upstream: https://github.com/riscv-software-src/opensbi |
+
+### Linux VM
+
+The `make run_linux` target will boot a bare Linux kernel as the host VM,
+though it doesn't do much without any filesystems or emulated devices.
+
+To launch a more functional Debian VM:
+- Grab pre-built `riscv64-virt` images here: https://people.debian.org/~gio/dqib/
+- Launch QEMU with:
+  ```
+  qemu-system-riscv64 \
+   -M virt,aia=aplic-imsic,aia-guests=4 \
+   -cpu rv64,x-aia=true -m 2048 -smp "${NCPU}" \
+   -nographic \
+   -bios "${OPENSBI}" \
+   -kernel "${SALUS}" -device guest-loader,addr=0xc0200000,kernel="${KERNEL}" \
+   -append "console=hvc0 earlycon=sbi keep_bootcon root=LABEL=rootfs" \
+   -device guest-loader,addr=0xc2200000,initrd="${INITRD}" \
+   -device x-riscv-iommu-pci \
+   -drive file="${ROOTFS}",if=none,id=hd \
+   -device nvme,serial=deadbeef,drive=hd \
+   -netdev user,id=usernet,hostfwd=tcp:127.0.0.1:7722-0.0.0.0:22 \
+   -device e1000e,netdev=usernet
```

Comment:
Can this replace the existing run_linux target in the makefile?

Reply:
This doesn't work for me with the following error.
panic : PanicInfo { payload: Any { .. }, message: Some(assertion failed: ram_size >= FDT_OFFSET + fdt_pages.length_bytes()), location: Location { file: "src/host_vm_loader.rs", line: 187, col: 9 }, can_unwind: true }

Where is the DTB located in the memory map ?

I had to change the memory to 4GB to make it work.

Reply:
Regarding run_linux target, can we have two options ?

1. run_linux_debian - that boots debian systems
2. run_linux - that boots just a busybox

The first one takes a long time to boot. So you many not want to test it always.

Reply:
> panic : PanicInfo { payload: Any { .. }, message: Some(assertion failed: ram_size >= FDT_OFFSET + fdt_pages.length_bytes()), location: Location { file: "src/host_vm_loader.rs", line: 187, col: 9 }, can_unwind: true }

Interesting, looks like this is a difference between the QEMU-bundled OpenSBI and the OpenSBI built from source: the former places the FDT at 0xbfe00000 and the latter places it right after the Salus image. This should be easy enough to fix; will send a PR.

Reply:
Added a `run_debian` target leaving `run_linux` unchanged.